lte/gateway/c/oai/tasks/s1ap/s1ap_mme_encoder.c

```diff
@@ -19,437 +19,145 @@
-    s1ap_message* message_p, uint8_t** buffer, uint32_t* length) {
-  S1ap_InitialContextSetupRequest_t initialContextSetupRequest;
-  S1ap_InitialContextSetupRequest_t* initialContextSetupRequest_p =
-      &initialContextSetupRequest;
-
-  memset(
-      initialContextSetupRequest_p, 0,
-      sizeof(S1ap_InitialContextSetupRequest_t));
+int s1ap_mme_encode_pdu(
+    S1ap_S1AP_PDU_t* pdu, uint8_t** buffer, uint32_t* length) {
+  int ret = -1;
 
-  if (s1ap_encode_s1ap_initialcontextsetuprequesties(
-          initialContextSetupRequest_p,
-          &message_p->msg.s1ap_InitialContextSetupRequestIEs) < 0) {
-    return -1;
+  if (pdu == NULL) {
+    OAILOG_DEBUG(LOG_S1AP, "PDU is NULL\n");
+  }
+  if (buffer == NULL) {
+    OAILOG_DEBUG(LOG_S1AP, "Buffer is NULL\n");
+  }
+  if (length == NULL) {
+    OAILOG_DEBUG(LOG_S1AP, "Length is NULL\n");
   }
 
-  return s1ap_generate_initiating_message(
-      buffer, length, S1ap_ProcedureCode_id_InitialContextSetup,
-      S1ap_Criticality_reject, &asn_DEF_S1ap_InitialContextSetupRequest,
-      initialContextSetupRequest_p);
-}
-
-//------------------------------------------------------------------------------
-int s1ap_mme_encode_pdu(
-    s1ap_message* message_p, uint8_t** buffer, uint32_t* length) {
-  DevAssert(message_p != NULL);
-  DevAssert(buffer != NULL);
-  DevAssert(length != NULL);
-
-  switch (message_p->direction) {
-    case S1AP_PDU_PR_initiatingMessage:
-      return s1ap_mme_encode_initiating(message_p, buffer, length);
+  switch (pdu->present) {
+    case S1ap_S1AP_PDU_PR_initiatingMessage:
+      ret = s1ap_mme_encode_initiating(pdu, buffer, length);
+      break;
 
-    case S1AP_PDU_PR_successfulOutcome:
-      return s1ap_mme_encode_successfull_outcome(message_p, buffer, length);
+    case S1ap_S1AP_PDU_PR_successfulOutcome:
+      ret = s1ap_mme_encode_successfull_outcome(pdu, buffer, length);
+      break;
 
-    case S1AP_PDU_PR_unsuccessfulOutcome:
-      return s1ap_mme_encode_unsuccessfull_outcome(message_p, buffer, length);
+    case S1ap_S1AP_PDU_PR_unsuccessfulOutcome:
+      ret = s1ap_mme_encode_unsuccessfull_outcome(pdu, buffer, length);
+      break;
 
     default:
       OAILOG_DEBUG(
           LOG_S1AP, "Unknown message outcome (%d) or not implemented",
-          (int) message_p->direction);
+          (int) pdu->present);
       break;
   }
-
-  return -1;
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_S1AP_PDU, pdu);
+  return ret;
 }
 
 //------------------------------------------------------------------------------
 static inline int s1ap_mme_encode_initiating(
-    s1ap_message* message_p, uint8_t** buffer, uint32_t* length) {
-  switch (message_p->procedureCode) {
-    case S1ap_ProcedureCode_id_downlinkNASTransport:
-      return s1ap_mme_encode_downlink_nas_transport(message_p, buffer, length);
+    S1ap_S1AP_PDU_t* pdu, uint8_t** buffer, uint32_t* length) {
+  asn_encode_to_new_buffer_result_t res = {NULL, {0, NULL, NULL}};
+  DevAssert(pdu != NULL);
```

Comment:
No need to assert here, just log an ERROR message and return. The entire process should not crash if one message is malformed.